tinygrad/lazy.py

```diff
@@ -257,8 +257,14 @@ def splitted_shape(dim_aft_div): return self.shape[:dim_to_split] + (self.shape[
   # *** movement ops ***
 
   def _movement_op(self, st: ShapeTracker, op: MovementOps, arg: Union[Tuple[sint, ...], Tuple[Tuple[sint, sint], ...]]) -> LazyBuffer:
-    if SHUFFLE_MOVEMENT_OPS and self.optype == BinaryOps and not self.realized and (op in {MovementOps.SHRINK, MovementOps.STRIDE, MovementOps.PERMUTE} or (op == MovementOps.RESHAPE and self.op.op in UnaryOps)) and not self.children:
-      return self.op.replace_with_movement_ops([(op, arg)])
+    if SHUFFLE_MOVEMENT_OPS and not self.realized and self.optype == BinaryOps and not self.children:
+      root_bufs = map(functools.partial(get_movementroot, allow_contiguous = True), self.op.buffers)
+      # don't push if all ast buffers (movement roots) are realized or sourceless
+      if not all(x.realized or len(x.op.src) == 0 for x in root_bufs):
+        push_expand_safe = all(x.realized or x.op.op not in LoadOps for x in root_bufs)
```

Comment:
When is pushing expands ever a good idea?

Reply:
expand doesn't allocate new memory right ?
![image](https://github.com/tinygrad/tinygrad/assets/88420255/e178bc11-4176-4cca-a814-4699aaa1c4d8)

the only point OOM was hitting me was when it was pushed above LoadOps. Hence the checking condition

Reply:
and coming onto the advantage. my understanding is pushing any movement op, so that BinOp comes next to each other (for fusion) is a good idea ?

Reply:
am getting test failures if i don't push
![image](https://github.com/tinygrad/tinygrad/assets/88420255/3b7d71db-0834-485b-9e7e-8cb135b4e5cc)

Reply:
That's likely a different issue.

Compare:
```
a = Tensor([[1,2,3]])
a = (a+1).expand(1000000,3)
```
with
```
a = Tensor([[1,2,3]])
a = a.expand(1000000,3)+1
```

In the bottom one we create a huge tensor, in the top we don't

Reply:
i actually get your point now. i was looking things from point of kernel fusion. if only # of kernels were important expand push would have really helped (see below example). But the memory & compute is getting messed up.

```
from tinygrad.tensor import Tensor

a = Tensor([[[[[1,2,3]]]]])
b = Tensor([[[[[1,2,3]]]]])

c = (a*b).expand((1,1,1,100,3))
d = Tensor([[[[[1,2,3]]]]])

e = (c+d).expand((1,1,100,100,3))
f = Tensor([[[[[1,2,3]]]]])

g = (f-e).expand((1,100,100,100,3))
h = Tensor([[[[[1,2,3]]]]])

i = (g*h).expand((100,100,100,100,3))
i.realize()
```

with expand push
![image](https://github.com/tinygrad/tinygrad/assets/88420255/76338437-df5d-48ad-9128-bc21d4fc2d67)

without expand push
![image](https://github.com/tinygrad/tinygrad/assets/88420255/e796f11b-7479-4f0c-be15-88b2dbbfe2d0)

Reply:
i will disable expand and see what is causing the tests to fail.